src/coreclr/jit/lsraxarch.cpp

```diff
@@ -3213,18 +3221,47 @@ int LinearScan::BuildMul(GenTree* tree)
         return BuildSimple(tree);
     }
 
-    // ToDo-APX : imul currently doesn't have rex2 support. So, cannot  use R16-R31.
-    int              srcCount      = BuildBinaryUses(tree->AsOp());
+    bool isUnsignedMultiply    = tree->IsUnsigned();
+    bool requiresOverflowCheck = tree->gtOverflowEx();
+    bool useMulx               = tree->OperGet() != GT_MUL && isUnsignedMultiply &&
+                   compiler->compOpportunisticallyDependsOn(InstructionSet_BMI2);
+
+    // ToDo-APX : imul currently doesn't have rex2 support. So, cannot use R16-R31.
+    int              srcCount      = 0;
     int              dstCount      = 1;
     SingleTypeRegSet dstCandidates = RBM_NONE;
 
-    bool isUnsignedMultiply    = ((tree->gtFlags & GTF_UNSIGNED) != 0);
-    bool requiresOverflowCheck = tree->gtOverflowEx();
+    // Lowering has ensured that op1 is never the memory operand
+    assert(!op1->isUsedFromMemory());
 
-    // There are three forms of x86 multiply:
+    // Start with building the uses, ensuring that one of the operands is in the implicit register (RAX or RDX)
+    // Place first operand in implicit register, unless:
+    //  * it is a memory address
+    // *  or the second operand is already in the register
+    if (useMulx)
+    {
+        // In lowering, we place any memory operand in op2 so we default to placing op1 in RDX
```

Comment:
For MULHI it seems like the constant is generally passed as op2, making the current implementation produce one more move than needed: 

for division by 10 it emits 
```
       mov      rax, 0xCCCCCCCCCCCCCCCD
       mov      rdx, rcx
       mulx     rax, rax, rax
       shr      rax, 3
```

but should be able to do

Reply:
It would be best to do this in lowering, the order that operands are built in should match the node's evaluation order, and that's something we generally do in lowering.